Source/WebCore/dom/FullscreenManager.cpp

```diff
@@ -301,84 +265,61 @@ void FullscreenManager::cancelFullscreen()
+        else
+            INFO_LOG(LOGIDENTIFIER, "Top document has no fullscreen element");
+    });
 }
 
-void FullscreenManager::exitFullscreen()
+static Vector<RefPtr<Document>> documentsToUnfullscreen(Document& doc)
 {
-    // The exitFullscreen() method must run these steps:
-
-    // 1. Let doc be the context object. (i.e. "this")
-    Document* currentDoc = &document();
-
-    // 2. If doc's fullscreen element stack is empty, terminate these steps.
-    if (m_fullscreenElementStack.isEmpty()) {
-        // If there is a pending fullscreen element but an empty fullscreen element stack,
-        // there is a pending task in requestFullscreenForElement(). Cause it to cancel and fire an error
-        // by clearing the pending fullscreen element.
-        INFO_LOG(LOGIDENTIFIER, "Cancelling pending fullscreen request.");
-        m_pendingFullscreenElement = nullptr;
-        return;
+    Vector<RefPtr<Document>> docs { RefPtr { &doc } };
+    while (true) {
+        auto lastDoc = docs.last();
+        if (!lastDoc)
+            continue;
+        ASSERT(lastDoc->fullscreenManager().fullscreenElement());
+        if (!lastDoc->fullscreenManager().isSimpleFullscreenDocument())
+            break;
+        auto parentFrame = lastDoc->frame()->ownerElement();
+        if (!parentFrame || (parentFrame && parentFrame->hasIframeFullscreenFlag()))
+            break;
+        docs.append(&parentFrame->document());
     }
+    return docs;
+}
 
+void FullscreenManager::exitFullscreen()
+{
     INFO_LOG(LOGIDENTIFIER);
 
-    // 3. Let descendants be all the doc's descendant browsing context's documents with a non-empty fullscreen
-    // element stack (if any), ordered so that the child of the doc is last and the document furthest
-    // away from the doc is first.
-    Deque<RefPtr<Document>> descendants;
-    for (AbstractFrame* descendant = frame() ? frame()->tree().traverseNext() : nullptr; descendant; descendant = descendant->tree().traverseNext()) {
-        auto* localFrame = dynamicDowncast<LocalFrame>(descendant);
-        if (!localFrame)
-            continue;
-        if (localFrame->document()->fullscreenManager().fullscreenElement())
-            descendants.prepend(localFrame->document());
-    }
+    auto* currentDoc = &document();
+    auto mode = ExitMode::NoResize;
+    auto exitDocs = documentsToUnfullscreen(*currentDoc);
 
-    // 4. For each descendant in descendants, empty descendant's fullscreen element stack, and queue a
-    // task to fire an event named fullscreenchange with its bubbles attribute set to true on descendant.
-    for (auto& document : descendants) {
-        document->fullscreenManager().clearFullscreenElementStack();
-        addDocumentToFullscreenChangeEventQueue(*document);
+    if (exitDocs.contains(&topDocument()) && topDocument().fullscreenManager().isSimpleFullscreenDocument()) {
+        mode = ExitMode::Resize;
+        currentDoc = &topDocument();
     }
```

Comment:
I think we should not call topDocument() three times. It’s not free to compute. So we can use a local variable.